### PR TITLE
DOC Add docs on napari models and events

### PR DESCRIPTION
**Pull request by lucyleeow**
# Description
Adds some basic docs about how napari models and events work. Happy to change anything misleading incorrect as I don't have an in-depth understanding of this yet.

It has been suggested to add a diagram, and I started with this simplistic example:
napari

Happy to take suggestions, not sure what to add/change.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

Relates to #1426

### CODE REVIEW

File: docs/guides/napari_models.md
Comment (Czaki) [inline docs/guides/napari_models.md]: The possible workaround for the problem of crashing the napari application on creating QAppliation is to set `QT_QPA_PLATFORM=offscreen` in the environment. 


obraz

I do not know if this will allow for proper render, for example, screenshots, but it should allow solving problems like #3686 or just make it possible to use on a server, the napari installation from conda.

Comment (lucyleeow) [reply docs/guides/napari_models.md]: Hmm I am not sure this would be the best place for this information, though admittedly there are no docs on how to run napari headless...

Comment (Czaki) [reply docs/guides/napari_models.md]: Also here you add information that import Qt could be problematic without explanation why.

Comment (lucyleeow) [reply docs/guides/napari_models.md]: Good point, I was mostly trying to summarise the info here: https://github.com/lucyleeow/napari/pull/7#discussion_r946160065

Note that I don't have a deep understanding of this. My guess would be if you are running in a CLI Qt may not work and crash?

Comment (Czaki) [reply docs/guides/napari_models.md]: The first command on sreenshot is the try of run napari (create viewer) on serwer without X. On try to create QApplication, if qt cannot connect to display then it print messages from terminal and abort application. And it will happen always on create viewer if qt libs are installed and default plugin is selected.

The environment variable told Qt backend that rendering should be done offscreen. So QApplication creating will end with success.

Comment (lucyleeow) [reply docs/guides/napari_models.md]: Thank you that makes sense. I think I will add this info to a separate section in the docs (tips on running napari headless) as this info may be useful to users, whereas the rest of this doc is more aimed at people interested in contributing to napari. Also it doesn't seem to fit with the content of these docs. Thanks

File: docs/guides/napari_models.md
Comment (lucyleeow) [inline docs/guides/napari_models.md]: Assumes #4955 has been merged

File: docs/guides/napari_models.md
Comment (Carreau) [inline docs/guides/napari_models.md]: I'm not sure which level you are targeting that for, but there is a number of implicit things in the black of code. There is a mix of standalone function and objects method. 
As `_update_display(...)` take self, then we can assume that this will be an object method.  Is mention  `self.dims.nsteps`, so I assume that `dims = Dims(...)` has m been assigned to the `self.dims` attribute, and thus I believe that `connect(_update_display)` should connect  to `self._update_display` somehow, otherwise the first argument that _update_displauy will get will be an `Event` with attributes like `source`, `value` ...

`nsteps` also don't exists on `Dims`.



```suggestion
dims = Dims(ndim=3, ndisplay=2)

# define some callback that should respond to changes in the model
# if the function takes a single parameter it will receive the event
# as first value. 
def _update_display(event):
    """
    Updates display for all sliders.
  
    """
    # the code updating the display code is not relevant for this
    # example thus has been omitted.
    
    # we can get the source object of the event, 
    assert event.source == dims 
    
    # ... and query any attributes
    ndisplay = event.source.ndisplay
    
    # ... or directly get the new value for this specific event:
    assert ndisplay == event.value
    print(f"Update number of dimensions displayed to {ndisplay}")

# register our callback with the model
dims.events.ndisplay.connect(_update_display)
```

File: docs/guides/napari_models.md
Comment (Carreau) [inline docs/guides/napari_models.md:159]: 👍

**Review by Carreau — APPROVED** [review]
+1, one suggestion to change one example, the additional thing I'll suggest is to remove the properties related to `ndim` as it is unused in the examples, and just leave it as an attribute called `ndim`

File: docs/guides/napari_models.md
Comment (Carreau) [inline docs/guides/napari_models.md]:
```suggestion
        self._ndim = ndim
```

Trailing underscore.

File: docs/guides/napari_models.md
Comment (lucyleeow) [reply docs/guides/napari_models.md]: > There is a mix of standalone function and objects method.

Thanks, mostly I was trying to give an easy example from the napari code base but I realise now that it was a method and did not make sense how I had it.

Comment (Carreau) [reply docs/guides/napari_models.md]: Thanks, the new version reads much better IMHO, thanks a lot for doing all that.

**Review by Carreau — APPROVED** [review]
All looks good to me, if no objection I'll merge in 24h or so, unless someone beats me to it.

**Comment by lucyleeow** [comment]
Maybe hold off on merging, I've had suggestions about adding a diagram (e.g., like https://github.com/napari/napari/pull/4929#issue-1337939922) but wasn't sure about my diagram/how useful it is. Would like to hear some feedback on this before merging if that is okay!

File: docs/guides/napari_models.md
Comment (kevinyamauchi) [inline docs/guides/napari_models.md:3]: Nice and clear explanation of the goal of the doc!

File: docs/guides/napari_models.md
Comment (kevinyamauchi) [inline docs/guides/napari_models.md]:
```suggestion
* python models describing components in the napari application - these are able to operate without the GUI
```

I think it might be nice to avoid "object" here to avoid confusion with programming concept.

**Review by kevinyamauchi — APPROVED** [review]
Hey @lucyleeow ! This looks great! I left some minor comments below.

With regards to the figure, I think it is a good starting point!  I think it might help clarify things to also have a couple of figures that show the flow of information for different actions. For example, one figure that shows the flow of information when something is updated in the GUI and another that shows the flow of information  Perhaps the arrows could be numbered so the order is clear.

That being said, this is already a huge improvement, so I would be happy to merge now and update the figure in a follow-up PR (whichever you prefer). What do you think?

File: docs/guides/napari_models.md
Comment (lucyleeow) [reply docs/guides/napari_models.md]: Thanks, this is better

**Comment by lucyleeow** [comment]
Thanks for the review!

> For example, one figure that shows the flow of information when something is updated in the GUI and another that shows the flow of information Perhaps the arrows could be numbered so the order is clear.

That's a good idea. Maybe a diagram in the 'qt classes' section showing what happens when something is updated in the GUI, and something in the 'python models and events' section showing what happens when something is updated in the console?

> That being said, this is already a huge improvement, so I would be happy to merge now and update the figure in a follow-up PR (whichever you prefer). What do you think?

I think it would be nice to get this in first and I can work on another PR for the figures (I need to work out lfs for this too :grimacing: )?

**Review by kevinyamauchi — APPROVED** [review]
> I think it would be nice to get this in first and I can work on another PR for the figures (I need to work out lfs for this too 😬 )?

Agreed!

This looks good to me! Thanks for the quick changes. I will merge now since it was sitting quite some time with the initial approval. Thanks @lucyleeow !

**Comment by lucyleeow** [comment]
Thanks!

**Pull request merged**
